### DISCUSSION TITLE
Issue with rust-analyzer LSP: helix tries to use the "wrong" channel

### DISCUSSION THREAD

**Discussion by jerabaul29** [helix-editor/helix#9226] [General]
I have helix installed:

```
✔ ~> hlx --version
helix 23.10 (f6021dd0)
```

and the rust-analyzer installed too:

```
✔ ~> hlx --health rust
Configured language servers:
  ✓ rust-analyzer: /home/jrmet/.cargo/bin/rust-analyzer
Configured debug adapter: lldb-vscode
Binary for debug adapter: 'lldb-vscode' not found in $PATH
Highlight queries: ✓
Textobject queries: ✓
Indent queries: ✓
✔ ~> /home/jrmet/.cargo/bin/rust-analyzer --version
rust-analyzer 1.73.0 (cc66ad4 2023-10-03)
```

However, it seems that helix does not like the channel for which the rust-analyzer is installed: when I open a project in the current project, the rust-analyzer lsp does not seem to work, and the ```log-open``` command gives me the reason in the logs:

```
2024-01-03T20:45:46.001 helix_lsp::transport [ERROR] rust-analyzer err <- "error: 'rust-analyzer' is not installed for the toolchain '1.65.0-x86_64-unknown-linux-gnu'\n"
2024-01-03T20:45:46.002 helix_lsp [ERROR] failed to initialize language server: server closed the stream
2024-01-03T20:45:46.002 helix_lsp::transport [ERROR] rust-analyzer err: <- StreamClosed
```

Checking my rustup, there is indeed such a version of the toolchain, but it is not my default:

```
✔ ~> rustup show
Default host: x86_64-unknown-linux-gnu
rustup home:  /home/jrmet/.rustup

installed toolchains
--------------------

stable-x86_64-unknown-linux-gnu (default)
nightly-x86_64-unknown-linux-gnu
1.65.0-x86_64-unknown-linux-gnu

installed targets for active toolchain
--------------------------------------

thumbv7em-none-eabihf
thumbv8m.main-none-eabihf
x86_64-unknown-linux-gnu

active toolchain
----------------

stable-x86_64-unknown-linux-gnu (default)
rustc 1.73.0 (cc66ad468 2023-10-03)
```

How should I make this work? :) Any way to force helix to use the "correctly installed and working" rust-analyzer version? Or should I install / maintain a number of different  versions of the rust-analyzer?

**Comment by jerabaul29** [comment]
Ok, I think that I understand: the project I was peeking into has a ```rust-toolchain.toml``` with the content:

```
[toolchain]
channel = "1.65.0"
components = ["rustfmt", "rust-src"]
```

So I guess this means that helix (or my terminal?) finds this file, and then forces rust tooling to use the 1.65.0? Is there a way / would it make sense to overwrite this behavior in helix somehow, so that I do not need to end up installing loads of versions of rust-analyzer that match every possible project I want to peek into?

**Comment by kirawi** [comment]
This is not a helix issue, this is just how rustup works. You can either install `rust-analyzer` outside of rustup or set `RUSTUP_TOOLCHAIN=<name>` which will apply to all Rust repositories.
